File: JourneyBot/Cogs/Feeds.py
import asyncio
import re
import datetime

import disnake
from disnake import ApplicationCommandInteraction
from disnake.ext import commands

import tortoise.exceptions

from Cogs.BaseCog import BaseCog
from Database.DBConnector import RedditFeed
from Views import Embed
from Util import Utils, Logging, Reddit
from Util.Emoji import msg_with_emoji


class Feeds(BaseCog):
    def __init__(self, bot: commands.Bot):
        super().__init__(bot)
        self.reddit_api = Reddit.get_reddit()
        # actually UUID, but str is good enough
        self.sleep_times: dict[str, int] = {}
        self.reported_fails: dict[str, bool] = {}
        self.tasks: dict[str, asyncio.Task] = {}
        self.sleep_start = 15
        self.sleep_stop = 7680

    async def cog_load(self):
        for feed in await RedditFeed.all():
            self.spawn_task(feed)

    def spawn_task(self, feed: RedditFeed):
        if self.sleep_times.get(feed.id) is None:
            self.sleep_times[feed.id] = 0
        if self.reported_fails.get(feed.id) is None:
            self.reported_fails[feed.id] = False
        task = self.bot.loop.create_task(self.update_reddit_feed(feed))
        self.tasks[feed.id] = task

    async def close(self):
        for task in self.tasks.values():
            task.cancel()
        timer = 0
        while timer <= 60:
            to_del = []
            for key, task in self.tasks.items():
                if task.done():
                    to_del.append(key)
            for item in to_del:
                del self.tasks[item]
            if len(self.tasks) == 0:
                break
            await asyncio.sleep(1)
            timer += 1

    @commands.slash_command(description="Feed management.")
    @commands.guild_only()
    @commands.default_member_permissions(ban_members=True)
    @commands.bot_has_permissions(send_messages=True)
    async def feed(self, inter: ApplicationCommandInteraction):
        pass

    @feed.sub_command(name="template-help", description="Feed template help.")
    async def template_help(self, inter: ApplicationCommandInteraction):
        embed = Embed.default_embed(
            title="Feed Help",
            description="Explanation of the template syntax.",
            author=inter.author.name,
            icon_url=inter.author.avatar.url,
        )
        embed.add_field(
            name="Line breaks",
            value="Line breaks are represented by `\\n`.",
            inline=False,
        )
        embed.add_field(
            name="Variables",
            value="Variables are replaced with the corresponding value from the feed.",
            inline=False,
        )
        embed.add_field(name="{{title}}", value="The title of the content.", inline=False)
        embed.add_field(name="{{link}}", value="The link to the content.", inline=False)
        await inter.response.send_message(embed=embed, ephemeral=True)

    @feed.sub_command(description="List all feeds in the server.")
    async def list(self, inter: ApplicationCommandInteraction):
        reddit_feeds = await RedditFeed.filter(guild=inter.guild_id).all()
        if not reddit_feeds:
            await inter.response.send_message("No feeds found.", ephemeral=True)
            return
        embed = Embed.default_embed(
            title="Feeds",
            description="List of all feeds in the server.",
            author=inter.author.name,
            icon_url=inter.author.avatar.url,
        )
        for feed in reddit_feeds:
            channel: disnake.abc.GuildChannel = Utils.coalesce(
                self.bot.get_channel(feed.channel),
                Utils.get_alternate_channel(feed.channel),
            )
            embed.add_field(
                name=f"#{channel.name} | ID: {feed.id}",
                value=f"r/{feed.subreddit}",
                inline=False,
            )
        await inter.response.send_message(embed=embed)

    @feed.sub_command_group(description="Add a feed to the server.")
    async def add(self, inter: ApplicationCommandInteraction):
        pass

    @add.sub_command(description="Add a reddit feed to the server.")
    async def reddit(
        self,
        inter: ApplicationCommandInteraction,
        subreddit_name: str = commands.Param(
            name="subreddit-name",
            description="The name of the subreddit.",
            min_length=1,
            max_length=21,
        ),
        template: str = commands.Param(default=None, name="template", description="The template for new posts."),
    ):
        regex = r"[a-zA-Z0-9_]{1,21}"
        if not re.match(regex, subreddit_name):
            await inter.response.send_message("Invalid subreddit name.", ephemeral=True)
            return
        if not template:
            template = "{{title}}\n{{link}}"
        else:
            template = template.replace("\\n", "\n")

        try:
            subreddit = await self.reddit_api.subreddit(subreddit_name, fetch=True)
            subreddit.id
        except Exception:
            await inter.response.send_message("Subreddit not found.", ephemeral=True)
            return
        feed = await RedditFeed.create(
            guild=inter.guild_id,
            channel=inter.channel_id,
            subreddit=subreddit_name,
            template=template,
        )
        self.spawn_task(feed)
        await inter.response.send_message(f"Added feed for r/{subreddit_name}.")
        await Logging.guild_log(
            inter.guild_id,
            msg_with_emoji(
                "FEED",
                f"A feed for r/{subreddit_name} (`{feed.id}`) was added to {inter.channel.mention} by {inter.author.name} (`{inter.author.id}`)",
            ),
        )
        Logging.info(
            f"A feed for r/{subreddit_name} ({feed.id}) was added to channel {inter.channel.name} ({inter.channel.guild.name}) by {inter.author.name} ({inter.author.id})"
        )

    @feed.sub_command(description="Remove a feed from the server.")
    async def remove(
        self,
        inter: ApplicationCommandInteraction,
        id: str = commands.Param(
            name="feed-id",
            description="The ID of the feed to remove.",
            min_length=36,
            max_length=36,
        ),
    ):
        try:
            feed = await RedditFeed.get(id=id, guild=inter.guild_id)
        except tortoise.exceptions.DoesNotExist:
            await inter.response.send_message("No feed found with that ID.", ephemeral=True)
            return
        channel: disnake.abc.GuildChannel = Utils.coalesce(
            self.bot.get_channel(feed.channel),
            Utils.get_alternate_channel(feed.channel),
        )
        await feed.delete()
        self.tasks[feed.id].cancel()
        del self.tasks[feed.id]
        await inter.response.send_message("Feed removed.")
        await Logging.guild_log(
            inter.guild_id,
            msg_with_emoji(
                "FEED",
                f"A feed for r/{feed.subreddit} (`{feed.id}`) was removed from {channel.mention} by {inter.author.name} (`{inter.author.id}`)",
            ),
        )
        Logging.info(
            f"A feed for r/{feed.subreddit} ({feed.id}) removed from channel {channel.name}"
            + f" ({inter.channel.guild.name}) by {inter.author.name} ({inter.author.id})"
        )

    @feed.sub_command(description="Manually restart a feed.")
    async def restart(
        self,
        inter: ApplicationCommandInteraction,
        id: str = commands.Param(
            name="feed-id",
            description="The ID of the feed to restart.",
            min_length=36,
            max_length=36,
        ),
    ):
        try:
            feed = await RedditFeed.get(id=id, guild=inter.guild_id)
        except tortoise.exceptions.DoesNotExist:
            await inter.response.send_message("No feed found with that ID.", ephemeral=True)
            return
        if feed.id not in self.restarts_available:
            await inter.response.send_message("This feed is not available for restart.", ephemeral=True)
            return
        self.restarts_available.remove(feed.id)
        self.bot.loop.create_task(self.update_reddit_feed(feed))
        await Logging.guild_log(
            feed.guild,
            msg_with_emoji(
                "FEED",
                f"A feed for {feed.subreddit} (`{feed.id}`) was manually restarted by {inter.author.name} (`{inter.author.id}`)",
            ),
        )
        Logging.info(f"Manually restarted feed {feed.id} ({feed.subreddit})")
        await inter.response.send_message("Attempting to restart feed.")

    async def update_reddit_feed(self, feed: RedditFeed):
        Logging.info(f"Starting feed {feed.id} ({feed.subreddit})")
        try:
            subreddit = await self.reddit_api.subreddit(feed.subreddit)
            async for submission in subreddit.stream.submissions():
                post_time = datetime.datetime.fromtimestamp(submission.created_utc, tz=datetime.timezone.utc)
                feed_latest_post = feed.latest_post
                if post_time > feed_latest_post:
                    await self.post_reddit_feed(feed, submission)
                    feed.latest_post = post_time
                    await feed.save()
            self.sleep_times[feed.id] = 0
            self.reported_fails[feed.id] = False
        except asyncio.CancelledError:
            Logging.info(f"Stopped feed for {feed.id} ({feed.subreddit})")
            return
        except Exception as e:
            Logging.warning(f"Exception in handle_post_error for {feed.id}: {e}")
            await self.handle_post_error(feed, e)

    async def post_reddit_feed(self, feed: RedditFeed, submission):
        channel = self.bot.get_channel(feed.channel)
        if not channel or not channel.permissions_for(channel.guild.me).send_messages:
            raise PermissionError
        await channel.trigger_typing()
        await asyncio.sleep(3)
        if not channel:
            Logging.guild_log(
                feed.guild,
                msg_with_emoji(
                    "WARN",
                    f"Unable to post to channel {feed.channel} for feed `{feed.id}` ({feed.subreddit})",
                ),
            )
            Logging.warning(f"Unable to post to channel {feed.channel} for feed {feed.id} ({feed.subreddit})")
            return
        message = feed.template.replace("{{title}}", submission.title)
        message = message.replace("{{link}}", f"https://www.reddit.com{submission.permalink}")
        await channel.send(message)

    async def handle_post_error(self, feed: RedditFeed, error: Exception):
        if isinstance(error, PermissionError):
            channel = self.bot.get_channel(feed.channel)
            c = f"channel `{feed.channel}`" if not channel else channel.mention
            await Logging.guild_log(
                feed.guild,
                msg_with_emoji(
                    "WARN",
                    f"I could not send a post for feed (`{feed.id}`) in {c}, because I don't have access to the channel. I will try again in 10 minutes.",
                ),
            )
            Logging.warning(f"Could not send post for feed {feed.id}. Channel {feed.channel} not found.")
            await asyncio.sleep(60 * 10)
        else:
            if self.sleep_times[feed.id] == 0:
                self.sleep_times[feed.id] = self.sleep_start
            await asyncio.sleep(self.sleep_times[feed.id])
            if self.sleep_times[feed.id] < self.sleep_stop:
                self.sleep_times[feed.id] *= 2
            else:
                if not self.reported_fails[feed.id]:
                    total_slept = 0
                    delay = self.sleep_start
                    while delay < self.sleep_stop:
                        total_slept += delay
                        delay *= 2

                    await Logging.guild_log(
                        feed.guild,
                        msg_with_emoji(
                            "WARN",
                            f"A feed for {feed.subreddit} (`{feed.id}`) has been failing for {int((total_slept / 60))} minutes."
                            + f" It will try to continue restarting every {int((self.sleep_stop / 60))} minutes.",
                        ),
                    )
                    Logging.error(
                        f"A feed for {feed.subreddit} ({feed.id}) has been failing for {int((total_slept / 60))} minutes."
                        + f" It will try to continue restarting every {int((self.sleep_stop / 60))} minutes."
                    )
                    self.reported_fails[feed.id] = True
        self.spawn_task(feed)


def setup(bot: commands.Bot):
    bot.add_cog(Feeds(bot))

File: journey-bot/src/commands/feeds.rs
use std::{sync::Arc, time::Duration};

use chrono::DateTime;
use poise::{
    CreateReply,
    serenity_prelude::{
        ChannelId, GuildId, Mentionable,
        futures::{self, Stream},
    },
};
use sea_orm::{
    ActiveModelTrait, ActiveValue::Set, ColumnTrait, EntityTrait, IntoActiveModel, QueryFilter,
};
use serde::{Deserialize, Deserializer};
use tokio_retry2::{Retry, RetryError, strategy::ExponentialFactorBackoff};
use tracing::{info, warn};

use crate::{
    Context, Error,
    emoji::Emoji,
    store::Store,
    utils::{BotError, LogError, eph, guild_log, send_message, timestamp_from_f64_with_tz},
    views::embed::default_embed,
};

#[cfg(feature = "reddit-api")]
use crate::{config::RedditConfig, utils::now};
#[cfg(feature = "reddit-api")]
use roux::{Me, Reddit, Subreddit, response::BasicThing, submission::SubmissionData};
#[cfg(feature = "reddit-api")]
use tokio::sync::RwLock;
#[cfg(feature = "reddit-api")]
use tracing::error;

#[derive(Deserialize, Clone, Debug)]
#[serde(rename = "feed")]
#[cfg(not(feature = "reddit-api"))]
pub struct Feed {
    #[serde(rename = "entry")]
    entries: Vec<Entry>,
}

#[derive(Deserialize, Clone, Debug)]
#[serde(rename = "entry")]
pub struct Entry {
    #[serde(deserialize_with = "deserialize_timestamp")]
    published: f64,
    title: String,
    #[serde(rename = "link")]
    link: Link,
}

#[derive(Deserialize, Clone, Debug)]
pub struct Link {
    #[serde(rename = "@href")]
    href: String,
}

fn deserialize_timestamp<'de, D>(deserializer: D) -> Result<f64, D::Error>
where
    D: Deserializer<'de>,
{
    let s = String::deserialize(deserializer)?;
    let dt = DateTime::parse_from_rfc3339(&s).map_err(serde::de::Error::custom)?;
    Ok(dt.timestamp() as f64)
}

#[cfg(feature = "reddit-api")]
impl From<BasicThing<SubmissionData>> for Entry {
    fn from(submission: BasicThing<SubmissionData>) -> Self {
        Self {
            published: submission.data.created_utc,
            title: submission.data.title,
            link: Link {
                href: format!("https://www.reddit.com{}", submission.data.permalink),
            },
        }
    }
}

#[derive(Clone, Debug, Deserialize)]
#[cfg(feature = "reddit-api")]
struct RedditToken {
    exp: f32,
}

#[cfg(feature = "reddit-api")]
pub struct RedditClient {
    client: RwLock<Option<Me>>,
    config: Option<RedditConfig>,
}

#[cfg(feature = "reddit-api")]
impl RedditClient {
    pub async fn new(config: Option<RedditConfig>) -> Result<Self, BotError> {
        if config.is_none() {
            warn!("Reddit config not provided, client will not be initialised");
        }
        let client = Self::make_client(&config).await;

        Ok(Self {
            client: RwLock::new(client),
            config,
        })
    }

    async fn make_client(config: &Option<RedditConfig>) -> Option<Me> {
        let Some(config) = config else {
            return None;
        };

        let client = Reddit::new(&config.user_agent, &config.id, &config.secret)
            .username(&config.username)
            .password(&config.password)
            .login()
            .await;

        match client {
            Ok(client) => Some(client),
            Err(e) => {
                error!("Failed to create reddit client: {e:?}");
                None
            }
        }
    }

    async fn get_client(&self) -> Result<Me, BotError> {
        let client = self.client.read().await;
        if let Some(client) = &*client {
            let token = client.config.access_token.clone().unwrap_or(String::new());
            let expiry = jsonwebtoken::dangerous::insecure_decode::<RedditToken>(token)
                .map(|t| t.claims.exp)
                .unwrap_or(0.0);
            let now = now().as_secs_f32();
            if expiry > (now - 60.0) {
                return Ok(client.clone());
            }
            info!("Rotating reddit access token")
        }
        drop(client);
        let mut client = self.client.write().await;
        *client = Self::make_client(&self.config).await;
        client
            .clone()
            .ok_or(BotError::new("Failed to create reddit client"))
    }

    fn can_init(&self) -> bool {
        self.config.is_some()
    }
}

pub struct RedditScheduler;

impl RedditScheduler {
    pub async fn schedule_all(store: Arc<Store>) -> Result<(), BotError> {
        for feed in sea_entity::reddit_feed::Entity::find()
            .all(&store.db.sea)
            .await?
        {
            Self::schedule(feed.id, store.clone());
        }
        Ok(())
    }

    pub fn schedule(id: String, store: Arc<Store>) {
        #[cfg(feature = "reddit-api")]
        if !store.reddit_client.can_init() {
            warn!(
                "Requested to schedule feed {}, but no reddit config is available",
                id
            );
            return;
        }
        tokio::spawn(async move {
            // sleep at startup so we have enough time to init everything
            tokio::time::sleep(Duration::from_secs(5)).await;
            Self::watch_subreddit(id, store.clone())
                .await
                .log("RedditScheduler::schedule")
        });
    }

    pub async fn watch_subreddit(id: String, store: Arc<Store>) -> Result<(), BotError> {
        info!("{}", format!("Started observer for feed {}", id));
        sea_entity::reddit_feed::Entity::find_by_id(&id)
            .one(&store.db.sea)
            .await?
            .ok_or(BotError::new("Feed not found"))?;

        async fn query_subreddit(
            #[allow(unused)] store: Arc<Store>,
            subreddit: &str,
        ) -> Result<Vec<Entry>, RetryError<()>> {
            #[cfg(feature = "reddit-api")]
            let entries = {
                let client = store
                    .reddit_client
                    .get_client()
                    .await
                    .map_err(|_| RetryError::to_transient::<()>(()).unwrap_err())?;

                let Some(subreddit_name) = subreddit.trim_end_matches("/").rsplit("/").next()
                else {
                    return RetryError::to_permanent(());
                };

                let subreddit = Subreddit::new_oauth(subreddit_name, &client.client);

                match subreddit.latest(25, None).await {
                    Ok(data) => Ok(data
                        .data
                        .children
                        .into_iter()
                        .rev()
                        .map(|e| e.into())
                        .collect()),
                    Err(err) => {
                        warn!("Got error from reddit api: {}", err);
                        RetryError::to_transient(())
                    }
                }
            };

            #[cfg(not(feature = "reddit-api"))]
            let entries = {
                let resp = match reqwest::get(subreddit).await {
                    Ok(resp) => resp,
                    Err(err) => {
                        if err.is_builder() {
                            return RetryError::to_permanent(());
                        }
                        return RetryError::to_transient(());
                    }
                };

                let text = resp
                    .error_for_status()
                    .map_err(|err| {
                        warn!("Error while fetching RSS document: {:?}", err);
                        RetryError::Transient {
                            err: (),
                            retry_after: None,
                        }
                    })?
                    .text()
                    .await
                    .map_err(|_| RetryError::Transient {
                        err: (),
                        retry_after: None,
                    })?;

                let feed = serde_xml_rs::from_str::<Feed>(&text).map_err(|err| {
                    warn!("Error deserializing XML document: {:?}", err);
                    RetryError::Transient {
                        err: (),
                        retry_after: None,
                    }
                })?;

                Ok(feed.entries)
            };

            entries
        }

        let retry_strategy = ExponentialFactorBackoff::from_millis(
            Duration::from_secs(15).as_millis().try_into().unwrap(),
            2.0,
        )
        .max_delay(Duration::from_secs(60 * 60 * 2));

        loop {
            // refetch every time, if it's removed we just terminate
            let Ok(Some(mut feed)) = sea_entity::reddit_feed::Entity::find_by_id(&id)
                .one(&store.db.sea)
                .await
            else {
                info!(
                    "{}",
                    format!(
                        "Terminated observer for feed {} because it doesn't exist anymore",
                        id
                    )
                );
                break Ok(());
            };

            let Ok(submissions) = Retry::spawn(retry_strategy.clone(), || {
                query_subreddit(store.clone(), &feed.subreddit)
            })
            .await
            else {
                guild_log(
                    store,
                    GuildId::new(feed.guild_id as u64),
                    Emoji::Warning,
                    format!("Tried to query entries for feed (`{}`), but URL was malformed. This will not be reattempted.", feed.id),
                    None,
                )
                .await;
                return Ok(());
            };

            let mut latest_time = feed.latest_post;
            for submission in submissions {
                if submission.published > latest_time {
                    let needs_retry = Self::post(store.clone(), &feed, &submission).await;
                    if needs_retry {
                        break;
                    }
                    latest_time = submission.published;
                    {
                        let mut updated_feed = feed.into_active_model();
                        updated_feed.latest_post = Set(submission.published);
                        feed = updated_feed.update(&store.db.sea).await?;
                    }
                }
            }

            let delay = if cfg!(feature = "reddit-api") {
                Duration::from_secs(60)
            } else {
                Duration::from_mins(10)
            };

            // we don't want to hammer the api
            tokio::time::sleep(delay).await;
        }
    }

    async fn post(
        store: Arc<Store>,
        feed: &sea_entity::reddit_feed::Model,
        submission: &Entry,
    ) -> bool {
        let msg = feed
            .template
            .replace("{{title}}", &submission.title)
            .replace("{{link}}", &submission.link.href);

        if let Err(err) = send_message(
            store.clone(),
            ChannelId::new(feed.channel_id as u64),
            msg,
            None,
        )
        .await
        {
            warn!(
                "Error while posting reddit post in guild {}: {err}",
                feed.guild_id
            );
            guild_log(
                store.clone(),
                GuildId::new(feed.guild_id as u64),
                Emoji::Warning,
                format!(
                    "I could not send a post message for feed (`{}`) in channel (`{}`). I will try again in 10 minutes.", feed.id, feed.channel_id
                ),
                None,
            ).await;
            tokio::time::sleep(Duration::from_secs(60 * 10)).await;
            true
        } else {
            false
        }
    }
}

#[poise::command(
    slash_command,
    subcommands("template_help", "list", "add", "edit", "remove"),
    guild_only,
    default_member_permissions = "BAN_MEMBERS",
    required_bot_permissions = "SEND_MESSAGES"
)]
pub async fn feed(_: Context<'_>) -> Result<(), Error> {
    Ok(())
}

/// Feed template help.
#[poise::command(slash_command, rename = "template-help")]
async fn template_help(ctx: Context<'_>) -> Result<(), Error> {
    let embed = default_embed(ctx)
        .title("Feed template help.")
        .description("Explanation of template syntax.")
        .fields(
            [
                ("Line breaks", "Line breaks are represented by `\\n`."),
                (
                    "Variables",
                    "Variables are replaced with the corresponding value from the feed.",
                ),
                ("{{title}}", "The title of the content."),
                ("{{link}}", "The link to the content."),
            ]
            .into_iter()
            .map(|(n, v)| (n, v, false)),
        );

    ctx.send(CreateReply::default().embed(embed).ephemeral(true))
        .await?;

    Ok(())
}

/// List all feeds in the server.
#[poise::command(slash_command)]
async fn list(ctx: Context<'_>) -> Result<(), Error> {
    let guild_id = ctx.guild_id().ok_or("Expected to be in a guild")?;

    let feeds = sea_entity::reddit_feed::Entity::find()
        .filter(sea_entity::reddit_feed::Column::GuildId.eq(guild_id.get()))
        .all(&ctx.data().db.sea)
        .await?;

    if feeds.is_empty() {
        eph(ctx, "No feeds found.").await?;
        return Ok(());
    }

    let mut embed = default_embed(ctx)
        .title("Feeds")
        .description("List of all feeds in the server.");

    for feed in feeds.into_iter() {
        let channel = ChannelId::new(feed.channel_id as u64)
            .name(ctx)
            .await
            .unwrap_or("Unknown".to_string());
        let latest_post =
            timestamp_from_f64_with_tz(feed.latest_post, ctx.data().clone(), guild_id).await;
        embed = embed.field(
            format!(
                "#{channel} | r/{} | ID: {} | LP: {}",
                feed.subreddit, feed.id, latest_post
            ),
            format!("`{}`", feed.template),
            false,
        );
    }

    ctx.send(CreateReply::default().embed(embed)).await?;

    Ok(())
}

#[poise::command(slash_command, subcommands("reddit"))]
async fn add(_: Context<'_>) -> Result<(), Error> {
    Ok(())
}

/// Add a reddit feed to the server.
#[poise::command(slash_command)]
async fn reddit(
    ctx: Context<'_>,
    #[rename = "subreddit-link"]
    #[description = "A link to the subreddit."]
    subreddit_link: String,
    #[description = "The template for new posts."] template: Option<String>,
) -> Result<(), Error> {
    let guild_id = ctx.guild_id().ok_or("Expected to be in a guild")?;

    ctx.defer().await?;

    #[cfg(feature = "reddit-api")]
    {
        let re = regex::Regex::new("[a-zA-Z0-9_]{1,21}").unwrap();

        let Some(partial_name) = subreddit_link.trim_end_matches("/").rsplit("/").next() else {
            eph(ctx, "Invalid subreddit link.").await?;
            return Ok(());
        };
        if !re.is_match(partial_name) {
            eph(ctx, "Invalid subreddit name.").await?;
            return Ok(());
        }

        let Ok(client) = ctx.data().reddit_client.get_client().await else {
            eph(ctx, "Reddit client is currently unavailable.").await?;
            return Ok(());
        };
        let subreddit = Subreddit::new_oauth(partial_name, &client.client);
        let Ok(_) = subreddit.about().await else {
            eph(ctx, "Subreddit not found.").await?;
            return Ok(());
        };
    }

    let template = template
        .unwrap_or("{{title}}\n{{link}}".to_string())
        .replace("\\n", "\n");

    let feed = sea_entity::reddit_feed::ActiveModel {
        id: Set(cuid2::slug()),
        guild_id: Set(guild_id.get() as i64),
        channel_id: Set(ctx.channel_id().get() as i64),
        subreddit: Set(subreddit_link.clone()),
        template: Set(template),
        ..Default::default()
    }
    .insert(&ctx.data().db.sea)
    .await?;

    ctx.say(format!("Added feed for `{}`", subreddit_link))
        .await?;

    RedditScheduler::schedule(feed.id.clone(), ctx.data().clone());

    guild_log(
        ctx.data().clone(),
        guild_id,
        Emoji::Feed,
        format!(
            "A feed `{}` (`{}`) was added to {} by {} (`{}`)",
            subreddit_link,
            feed.id,
            ctx.channel_id().mention(),
            ctx.author().name,
            ctx.author().id
        ),
        None,
    )
    .await;

    Ok(())
}

async fn autocomplete_id<'a>(
    ctx: Context<'_>,
    partial: &'a str,
) -> impl Stream<Item = String> + 'a {
    let guild_id = ctx.guild_id().unwrap_or(GuildId::new(1));
    let feeds = sea_entity::reddit_feed::Entity::find()
        .filter(sea_entity::reddit_feed::Column::GuildId.eq(guild_id.get()))
        .all(&ctx.data().db.sea)
        .await
        .unwrap_or(Vec::new());

    futures::stream::iter(
        feeds
            .into_iter()
            .filter(move |m| m.id.starts_with(partial))
            .map(|m| m.id),
    )
}

/// Edit a feed.
#[poise::command(slash_command)]
async fn edit(
    ctx: Context<'_>,
    #[description = "The ID of the feed to edit."]
    #[min_length = 10]
    #[max_length = 10]
    #[autocomplete = "autocomplete_id"]
    id: String,
    #[description = "The new template to use for the feed"] template: String,
) -> Result<(), Error> {
    let guild_id = ctx.guild_id().ok_or("Expected to be in a guild")?;

    let feed = sea_entity::reddit_feed::Entity::find_by_id(&id)
        .filter(sea_entity::reddit_feed::Column::GuildId.eq(guild_id.get()))
        .one(&ctx.data().db.sea)
        .await?;

    let Some(feed) = feed else {
        eph(ctx, "No feed found with that ID.").await?;
        return Ok(());
    };

    let mut feed = feed.into_active_model();
    feed.template = Set(template.replace("\\n", "\n"));
    feed.update(&ctx.data().db.sea).await?;

    guild_log(
        ctx.data().clone(),
        guild_id,
        Emoji::Feed,
        format!(
            "Feed `{}` was updated by {} (`{}`)",
            id,
            ctx.author().name,
            ctx.author().id
        ),
        None,
    )
    .await;

    ctx.say("Feed updated.").await?;

    Ok(())
}

/// Remove a feed from the server.
#[poise::command(slash_command)]
async fn remove(
    ctx: Context<'_>,
    #[description = "The ID of the feed to remove."]
    #[min_length = 10]
    #[max_length = 10]
    #[autocomplete = "autocomplete_id"]
    id: String,
) -> Result<(), Error> {
    let guild_id = ctx.guild_id().ok_or("Expected to be in a guild")?;

    let feed = sea_entity::reddit_feed::Entity::find_by_id(&id)
        .filter(sea_entity::reddit_feed::Column::GuildId.eq(guild_id.get()))
        .one(&ctx.data().db.sea)
        .await?;

    let Some(feed) = feed else {
        eph(ctx, "No feed found with that ID.").await?;
        return Ok(());
    };

    let channel = ChannelId::new(feed.channel_id as u64);

    let guild_log_msg = format!(
        "A feed for r/{} (`{}`) was removed from {} by {} (`{}`)",
        feed.subreddit,
        feed.id,
        channel.mention(),
        ctx.author().name,
        ctx.author().id
    );

    feed.into_active_model().delete(&ctx.data().db.sea).await?;

    ctx.say("Feed removed.").await?;

    guild_log(
        ctx.data().clone(),
        guild_id,
        Emoji::Feed,
        guild_log_msg,
        None,
    )
    .await;

    Ok(())
}
